token_metadata: read remapping for write_both_read_new

gusev-p opened the pull request:
When new nodes are added or existing nodes are deleted, the topology state machine needs to shunt reads from the old nodes to the new ones. This happens in the `write_both_read_new` state. The problem is that previously this state was not handled in any way in `token_metadata` and the read nodes were only changed when the topology sm reached the final 'owned' state.

To handle `write_both_read_new` an additional `interval_map` inside `token_metadata` is maintained similar to `pending_endpoints`.  It maps the ranges affected by the ongoing topology change operation to replicas which should be used for reading. When topology state sm reaches the point when it needs to switch reads to a new topology, it passes `request_read_new=true` in a call to `update_pending_ranges`. This forces `update_pending_ranges` to compute the ranges based on new topology and store them to the `interval_map`. On the data plane, when a read on coordinator needs to decide which endpoints to use, it first consults this `interval_map` in `token_metadata`, and only if it doesn't contain a range for current token it uses normal endpoints from `effective_replication_map`.

gusev-p commented:
This is more of RFC, I'm not sure about the approach. Also, we need some unit tests for this.

gleb-cloudius commented:
Current code calculates `_pending_ranges_interval_map`  which contains the additional replicas writes should go to per range per keyspase. What prevents us from re-using it for reads is that unlike writes we need to know what replicas are losing the  range, while the data structure only contain information about who is gaining it. May be we can extend `_pending_ranges_interval_map` with remapping information.

gusev-p commented:
1. The approach in this PR relies on the observation that we don't need to bother with 'non-primary' replicas - those that are added by replication strategies to the main one, derived from the ring - `_token_to_endpoint_map`. This is the case, since if we add an X node onto a ring between two existing nodes A and B (normal nodes), then it splits (A, B] to (A, X] and (X, B] and all replicas of the original range will still apply to the new range (A, B]. The only thing that we need to correct for reads, is that if the read is happen to be in (A, X] then we need to redirect in from B to X.
2. Actually, the same logic seems true for writes - if write happens in (A, X] then we only need to add X to the list of normal nodes, which includes B. I don't understand why we need all the tricky logic in `update_pending_ranges`, which builds all the ranges with regard to replication strategies. It's quite possible that I didn't get some nuances, that's why I said "I'm not sure about the approach" above. If this approach if correct, we can replace `_pending_ranges` with this structure which would serve both reads and writes.
3. `Also, the approach here seems limiting, because it remaps one replica. IIUC, it assumes that at most one node is leaving or joining so one replica is remapped.` No, there is no such assumption. Suppose we are adding two nodes X1 and X2 between normal nodes A and B. If the read is from (A, X1], then we need to remap B->X1, if it's from (X1, X2] then we need to remap B->X2. There is no need to remap many nodes at once.

kbr-scylla commented:
> Suppose we are adding two nodes X1 and X2 between normal nodes A and B. If the read is from (A, X1], then we need to remap B->X1, if it's from (X1, X2] then we need to remap B->X2. There is no need to remap many nodes at once.

I don't get it. By `(A, X1]` you mean range between two tokens, one owned by `A` and one owned by `X1`?

Let's say we had the following tokens in the token ring, owned by the following nodes: `100: A, 400: B, 500: C`. Now we add two more tokens owned by: `200: X1, 300: X2`. Let's say we have a keyspace with SimpleStrategy RF=3. When you read from token `50`: before the addition, the replica set of `50` in this keyspace is `A, B, C`, after the addition, the replica set of `50` in this keyspace is `A, X1, X2`. So there are two different replicas you need to read from in `write_both_read_new`.

kbr-scylla commented:
> I don't understand why we need all the tricky logic in update_pending_ranges, which builds all the ranges with regard to replication strategies.

It might be because the calculations there support, in theory, multiple nodes joining/leaving at the same time. E.g. there could be two pending replicas. It's weird because we actually never correctly supported booting multiple nodes at the same time. I don't know if Cassandra ever did, but I doubt it. Maybe they planned to, wrote this generic logic for replica set calculations, then gave up at some point.

gusev-p commented:
> Let's say we had the following tokens in the token ring, owned by the following nodes: `100: A, 400: B, 500: C`. Now we add two more tokens owned by: `200: X1, 300: X2`. Let's say we have a keyspace with SimpleStrategy RF=3. When you read from token `50`: before the addition, the replica set of `50` in this keyspace is `A, B, C`, after the addition, the replica set of `50` in this keyspace is `A, X1, X2`. So there are two different replicas you need to read from in `write_both_read_new`.

That's what I got wrong) `10->A,20->X,30->B,40->C,50->D, SimpleStrategy, RF=3, A, B, C, D` - existing nodes, X-new node. Remapping `B->X` is not sufficient to handle `write_both_read_new` for reads, we also need to stop using D. That's why we need all this fuss with replication strategies.

OK, will add something similar to update pending ranges.

kbr-scylla commented:
> 10->A,20->X,30->B,40->C,50->D, SimpleStrategy, RF=3, A, B, C, D - existing nodes, X-new node. Remapping B->X is not sufficient to handle write_both_read_new for reads, we also need to stop using D.

Assuming the tokens you listed are the only tokens in the ring:
1. For tokens in the range (40, 50], the replica set changes from D,A,B to D,A,X
2. For tokens in the range (50, 10] (wrap-around the ring): from A,B,C to A,X,B
3. Range (10, 20]: B,C,D to X,B,C
4. (20, 30]: B,C,D to B,C,D (unchanged)
5. (30, 40]: C,D,A to C,D,A (unchanged)

Each case where the replica set changes can be expressed as a remapping, although I'm not sure if this is the simplest way to look at this:
1. B -> X
2. C -> X
3. D -> X

it seems like using remappings is sufficient, at least when only one node is added / removed, but calculating the remmapings might be nontrivial.

gleb-cloudius commented:
We rely on the fact that if only one node is added at a time there can only be one pending node. The trick is to calculate which one is it.

tgrabiec commented:
The calculation can be made simple if we structure the code like in the "arbitrary changes" spec. That is, have two rings, old and new. Calculate replicas for the old ring, and for the new ring. That gives you for every token two sets of replicas, old and new. pending = new - old. Even if we don't plan to support arbitrary movements, that makes the logic simple.

gusev-p commented:
v2:
  * implemented @tgrabiec's idea - refactored `update_pending_ranges` to build new version of `token_metadata` and compute `pending` with `new - old`

gusev-p commented:
While working on this PR a question begged: why do `pending_ranges` live in `token_metadata` and not in `effective_replication_map`?
* `effective_replication_map` is the final immutable representation of how replication should work on the ring, `pending_ranges` seem logically closer to it;
* `pending_ranges` in `effective_replication_map` would allow `effective_replication_map::factory_key` to be used as a key, not the whole keyspace name; different keyspaces can share the same `effective_replication_map`, reducing some overhead;
* we can remove `update_pending_ranges` altogether.

@bhalevy @gleb-cloudius

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/491/

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -370,6 +370,9 @@ future<> storage_service::topology_state_load() {
                 tmptr->update_topology(ip, locator::endpoint_dc_rack{rs.datacenter, rs.rack});
                 tmptr->add_bootstrap_tokens(rs.ring.value().tokens, ip);
+                if (rs.ring.value().state == ring_slice::replication_state::write_both_read_new) {
+                    tmptr->request_read_new(ip);
```
We desperately need tests.

gusev-p replied:
We can check the result of `calculate_natural_endpoints`, if `new_endpoints` (node added) or `old_endpoints` (node removed) contains a `request_read_new`-node - we add to `read_remapping`.

It posses another problem though - now we consider all the changes to the ring combined, some of them maybe marked with `request_read_new`, some not. It seems we should consider two versions of `new_token_metadata` - one with all changes (for `write_remapping`), and another with only `request_read_new`-changes (for `read_remapping`).

gusev-p commented:
v3:
  * comments added
  * ring_entries removed
  * fixed comment for update_pending_ranges

things to be done in this PR:
* unit tests for token_metadata (`update_pending_ranges`, `pending_endpoints_for`, `remapped_endpoints_for_reading`)
* `read_remapping` bug 1: `request_read_new` is not accounted for non-primary replicas
* `read_remapping` bug 2: we should ignore non `request_read_new`-changes when building `read_remapping`
* `request_read_new` should be called for replace, as long as for add and remove node

ideas (not yet decided on) for the follow-up PRs:
* split `token_metadata` to mutable builder part and immutable runtime part;
* move `update_pending_ranges` to `effective_replication_map`;
* build `new_token_metadata` once for all keyspaces;
* extract `interval_to_range` from `token_metadata`, it seems strange there.

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/887/

kbr-scylla commented on locator/token_metadata.cc:
```diff
@@ -769,122 +719,113 @@ token_metadata_impl::has_pending_ranges(sstring keyspace_name, inet_address endp
+            {
+                std::unordered_map<inet_address, std::unordered_set<token>> bootstrap_tokens;
+                for (const auto& [token, inet_address]: _bootstrap_tokens) {
+                    bootstrap_tokens[inet_address].insert(token);
```
> replacing e1 with e2 and bootstrap e2 at the same time

This is the only possible way to replace. You cannot perform a replace without bootstrapping a node.

But what's the "collision" in this scenario? The tokens that the replacing node will own are exactly the tokens that the replaced node has owned. There are no additional tokens introduced. `_bootstrap_tokens` should be empty in that case and if not, they should be exactly the same as what you find in the previous loop.

Anyway what I meant is that there must not be two different nodes e2 and e3 such that they are introducing the same tokens (either by replacing or by bootstrapping).

gusev-p replied:
> But what's the "collision" in this scenario?

```
replace(e1 -> e2)
bootstrap(e2, {e2_tokens})
```

The conflict is in what tokens should be assigned to e2, those that was initially assigned to e1, or e2_tokens.

Maybe my confusion is that I don't fully understand why we need replace in the first place and what are its usage scenarios.

kbr-scylla replied:
>  The conflict is in what tokens should be assigned to e2, those that was initially assigned to e1, or e2_tokens.

`e2` does not introduce any tokens on its own during replace, it takes `e1`'s tokens. There is no way to both introduce new tokens and steal someone else's tokens currently. If there was we would have to take the sum of both when calculating `new_normal_tokens`.

kbr-scylla replied:
> Maybe my confusion is that I don't fully understand why we need replace in the first place and what are its usage scenarios.

It's faster to perform a replace operation than to `removenode` and then follow-up with bootstrap.

gusev-p replied:
1. The issue seems not particularly important, since these conflicts can't occur in practice, if I understand correctly. We need some consistent behaviour with min scope of change.
2. From the `token_metadata` point of view we can treat `replace(e1, e2)` just as a shortcut for 
`bootstrap(e2, e1_tokens); leave(e1)`;
3. Currently we can call `bootstrap(e, <some tokens>)` many times, the effect is that the tokens are merged.

So it seems ok to just merge tokens in our case too. This means I can drop `bootstrap_tokens` and update `new_normal_tokens` directly, as you proposed.

kbr-scylla commented on locator/token_metadata.cc:
```diff
@@ -769,122 +719,113 @@ token_metadata_impl::has_pending_ranges(sstring keyspace_name, inet_address endp
+            ring_entries.insert(ring_entries.end(), bootstrap_tokens.begin(), bootstrap_tokens.end());
+            std::sort(begin(ring_entries), end(ring_entries));
+            return ring_entries;
+        }();
```
> What data structure do you mean? Any balanced tree will have bad data locality)

Maybe a B tree/B+ tree. AFAIR we introduced B+ trees in memtables to improve cache-friendliness among other things (previously they used red-black trees)

kbr-scylla replied:
> we still need ring_entries for leaving endpoints, returned it back

Catched by test? :D

gusev-p replied:
> Catched by test? :D

Yes :)

kbr-scylla commented on locator/token_metadata.cc:
```diff
@@ -769,122 +719,113 @@ token_metadata_impl::has_pending_ranges(sstring keyspace_name, inet_address endp
-    all_left_metadata->_impl->sort_tokens();
-}
-
 future<> token_metadata_impl::update_pending_ranges(
```
> total: O((n + m) * (replication_factor + log(n + m))).

The `n*m` from first bullet is missing.

Although I don't understand where the `n*m` comes from here:
> new_token_metadata: O(n * m + (n + m)log(n+m)), multiplication is from inefficient linear scans in update_normal_tokens, log is from sort_tokens

IIUC `update_normal_tokens` itself is `(n+m)log(n+m)`. Are you multiplying by `n` because of this loop?
```
            // apply new_normal_tokens
            for (auto& [endpoint, tokens]: new_normal_tokens) {
                result->update_topology(endpoint,
                    get_dc_rack(endpoint),
                    _normal_token_owners.contains(endpoint) ? node::state::normal : node::state::joining);
                result->update_normal_tokens(std::move(tokens), endpoint).get();
            }
```
but (at least today) there should be at most one pair in `new_normal_tokens`, so there should be a single call to `update_normal_tokens` with `tokens.size() == n`. So it should be just `(n+m)log(n+m)` from the single call to `update_normal_tokens`.

If the new tokens were spread among multiple endpoints the complexity analysis seems more complex. If we naively estimated `new_normal_tokens.size() = O(n)` then we would get `O(n*(n+m)log(n+m))`.

gusev-p replied:
I don't feel much value in precise O-estimation of the two approaches.

The important thing (mentioned in the refactoring commit message) is that old approach called `get_ranges` for each affected node. `get_ranges` is basically `calculate_natural_endpoints` for each token. So, old approach did that at least once, new approach - exactly only -> it's not worse than the old.

kbr-scylla commented on locator/token_metadata.cc:
```diff
@@ -59,7 +59,17 @@ class token_metadata_impl final {
+    // This data structure is filled only during data migration between nodes
+    // when they are added or removed from the cluster.
+    // During normal operation, token mapping to nodes is
+    // implemented in the effective_replication_map.
```
The comment suggests that the structure stores exactly the replicas to which we'll write:
```
     ... migration_info contains ranges of tokens and
    // corresponding replicas to which writes or reads will be directed.
```
But the truth is that `write_remapping` only contains the pending replicas, which are added to the natural replicas (obtained from effective_replication_map) during writes.

I still think we should not use the name "remapping" for writes because it's confusing, it suggests that we replace the usual replica set with something completely different, while we're summing both. `pending_replicas` or `pending_ranges` or `additional_replicas` are all appropriate.

gusev-p commented:
v4:

* add unit tests for `pending_endpoints_for` and `endpoints_for_reading`;
* fixed `read_remapping`, `request_read_new` sets global flag which is not bound to any endpoints;
* `request_read_new` called for replace nodes.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/905/

bhalevy commented on locator/token_metadata.cc:
```diff
@@ -763,122 +713,108 @@ token_metadata_impl::has_pending_ranges(sstring keyspace_name, inet_address endp
+            result.reserve(get_token_to_endpoint().size() + get_bootstrap_tokens().size());
+            for (const auto& p: get_token_to_endpoint()) {
+                result.push_back(p.first);
+            }
```
Why not simply copy `sorted_tokens()`?

gusev-p replied:
Is it simpler? we still need bootstrap tokens, similar looking loops read easier to me.

bhalevy replied:
sorted_token can be copied in bulk using std::copy soit's both simpler and more efficient.

gusev-p replied:
ok, will update

kbr-scylla replied:
If we care about efficiency like this, the following algorithm would be appropriate:
0. reserve space in `result` vector for all tokens
1. copy `sorted_tokens()` to the first part of `result` vector
2. copy `bootstrap_tokens` to the second part of `result` vector
3. sort only the second part
4. `std::merge` the two parts

gusev-p replied:
> If we care about efficiency like this

I definitely don't, the complexity/utility ratio is just too high.

kbr-scylla commented on locator/token_metadata.cc:
```diff
@@ -763,122 +713,108 @@ token_metadata_impl::has_pending_ranges(sstring keyspace_name, inet_address endp
+            // construct new_normal_tokens based on _bootstrap_tokens and _replacing_endpoints
+            std::unordered_map<inet_address, std::unordered_set<token>> new_normal_tokens;
+            if (!_replacing_endpoints.empty()) {
+                for (const auto& [token, inet_address]: _token_to_endpoint_map) {
```
I feel like we're missing some important `maybe_yield()`s. There were a bunch of those in the old code, this code seems to have fewer of them.

To be safe, maybe let's add a `maybe_yield()` whenever we're iterating over tokens: here, below for `_bootstrap_tokens` loop, then `new_normal_tokens` loop.

Curious that `sort_tokens()` doesn't yield though. Maybe it's not necessary. The loops are almost linear (with `log` here and there). Still, the yield shouldn't hurt (cc @bhalevy - please comment)

gusev-p replied:
I was also curios about those `maybe_yield`s:
  * why we have so many of them in the `token_metadata` code? Most of the loops are linear, number of elements should be small, they all should be taking nanoseconds to execute.
  * what is the cost of them? `need_preempt` is `atomic_signal_fence(std::memory_order_seq_cst)` and two relaxed reads, I'm not sure if this heavier than the iterations of the loops.

kbr-scylla replied:
> why we have so many of them in the token_metadata code? Most of the loops are linear, number of elements should be small, they all should be taking like nanoseconds to execute.

Probably because these smaller functions are called from larger functions which sometimes have quadratic complexity etc. These yields were being added over time, solving reactor stalls.

> what is the cost of them? need_preempt is atomic_signal_fence(std::memory_order_seq_cst) and two relaxed reads, I'm not sure if this is heavier than the iterations of the loops.

Note that `maybe_yield` doesn't yield unconditionally, hence the `maybe_`. I don't know about the cost of yield itself though. But it should be insignificant.

gusev-p replied:
> Note that maybe_yield doesn't yield unconditionally, hence the maybe_. 

Yes, but it uses `need_preempt` to decide whether or not to yield.

> Probably because these smaller functions are called from larger functions which sometimes have quadratic complexity etc.

This means we can have just one `maybe_yield` per function, not per loop iteration.

bhalevy replied:
The respective patches should provide the context for why the maybe_yield calls were added. Basically this was in response to stalls seen with a large number of tokens.
As for sorting the tokens I actually had a branch with a gentle sorting algorithm that would consider yielding in the loop instead of using std::sort.  IIRC it required a very large number of tokens to reproduce a significant stall when sorting that I thought it's not worth submitting it, but maybe we should reconsider it now that we have large enough clusters out there.

Anyhow, anything that has quadratic complexity or worse of |tokens| needs yielding. Linear should be fine, O(n*log(n)) is marginal, depnding on n and the constant multiplier.

gusev-p replied:
Besides sorting tokens, the only part of the code that is above linear is the loop to compute `migration_info`, and we already have `thread::maybe_yield` there, so no more `maybe_yield` is needed.

kbr-scylla commented on locator/token_metadata.cc:
```diff
@@ -830,6 +838,12 @@ future<> token_metadata_impl::update_pending_ranges(
+
+                // in order not to waste memory, we update read_endpoints only if the
+                // new endpoints differs from the old one
+                if (_read_new_requested && new_endpoints.get_vector() != old_endpoints.get_vector()) {
```
Maybe we should compare sets for safety instead of vectors - since comparing vectors depends on the order of elements.

However since this is only an optimization, it doesn't matter that much.

gusev-p replied:
Comparing vectors is more lightweight. They are populated by the same function, the order must be the same.

kbr-scylla commented on test/boost/token_metadata_test.cc:
```diff
@@ -189,3 +189,59 @@ SEASTAR_THREAD_TEST_CASE(test_pending_endpoints_for_replace_with_replicas) {
         inet_address_vector_topology_change{});
 }
+
+SEASTAR_THREAD_TEST_CASE(test_endpoints_for_reading_when_bootstrap_with_replicas) {
```
A test for replace and leaving would be great as well, but can be left for follow-up.

kbr-scylla reviewed (approved):
Thanks for the tests.

I think the code is correct now, but @tgrabiec @bhalevy please review as well. This is a significant change which affects like 90% of the system (write/read path replica set calculation). We need multiple eyes on this.

gusev-p commented:
v5:
  * rebased on master (with tablets);
  * use `sorted_tokens()` for tokens calculation in `update_pending_ranges`;
  * update `update_pending_ranges` comment.

gusev-p commented:
v6:
  * simple merge fix

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/951/

bhalevy commented on locator/token_metadata.cc:
```diff
@@ -773,122 +710,107 @@ token_metadata_impl::has_pending_ranges(sstring keyspace_name, inet_address endp
     if (_bootstrap_tokens.empty() && _leaving_endpoints.empty() && _replacing_endpoints.empty()) {
         tlogger.debug("No bootstrapping, leaving nodes, replacing nodes -> empty pending ranges for {}", keyspace_name);
-        set_pending_ranges(keyspace_name, std::unordered_multimap<range<token>, inet_address>(), can_yield::no);
+        _pending_ranges_interval_map.erase(keyspace_name);
```
for followup: consider clear_gently here.

gusev-p replied:
ok

bhalevy commented on locator/token_metadata.cc:
```diff
@@ -773,122 +710,107 @@ token_metadata_impl::has_pending_ranges(sstring keyspace_name, inet_address endp
+            auto tokens = std::vector<dht::token>();
+            tokens.reserve(sorted_tokens().size() + get_bootstrap_tokens().size());
+            tokens.resize(sorted_tokens().size());
+            std::copy(begin(sorted_tokens()), end(sorted_tokens()), begin(tokens));
```
why is `resize` needed?

typically, this is done as `std::copy(sorted_tokens().begin(), sorted_tokens().end(), std::back_inserter(tokens))`
does that prevent bulk copying (`dht::token` should have a trivial copy constructor, no?)

gusev-p replied:
```
-O3

resize -> memmove
back_inserter -> __copy_move<false, false, ...> -> copy item by item
```

https://godbolt.org/z/7Tz7E4Mqv

bhalevy replied:
Hmm, so this could probably be the most efficient way:
```c++
    auto tokens = sorted_tokens();
    tokens.reserve(tokens().size() + get_bootstrap_tokens().size());
    boost::range::copy(get_bootstrap_tokens() | boost::adaptors::map_keys, std::back_inserter(tokens));
```

gusev-p replied:
I don't like this boost magic, the variant with loop is simple and straightforward.

And it's difficult to decipher what's going on in the assembly https://godbolt.org/z/xa1TKnE6j. I doubt it's more efficient than the simple loop, in either way we need to collect the values from different memory locations/hashtable buckets.

gleb-cloudius replied:
I do not insist of course. You are not the only one who dislikes boost (I do not mind it), but you do use `std::copy()` above instead of an open loop (which is good). Using an algorithm is better than open code a loop because it clearly shows the intention of the code without a need to read the loop. IIRC we do not use c++20 ranges because clang does not support it this is why we have to use boost sometimes.

bhalevy replied:
ok, this should be too frequently called anyway, so we don't need to squeeze every cpu cycle out of it.

bhalevy commented on locator/token_metadata.cc:
```diff
@@ -903,6 +921,14 @@ inet_address_vector_topology_change token_metadata_impl::pending_endpoints_for(c
+std::optional<inet_address_vector_replica_set> token_metadata_impl::endpoints_for_reading(const token& token, const sstring& keyspace_name) const {
+    const auto* endpoints = maybe_migration_endpoints(endpoints_field::read_endpoints, token, keyspace_name);
+    if (endpoints == nullptr) {
+        return {};
```
better return std::nullopt;

gusev-p replied:
why is it better?

bhalevy replied:
When returning `nullopt` it's clear that the function is returning a disengaged optional.
Otherwise one needs to look at the return type to figure what is returned.  The other return path returns a vector so this one could be mistaken to return an empty vector.

gusev-p replied:
ok

gusev-p commented on locator/token_metadata.cc:
```diff
@@ -176,6 +178,8 @@ class token_metadata_impl final {
     void add_leaving_endpoint(inet_address endpoint);
     void del_leaving_endpoint(inet_address endpoint);
+
+    void request_read_new();
```
in the follow-up https://github.com/scylladb/scylladb/pull/13715 we are going to move `pending_ranges` to erm, and we would need this a flag in token_metadata, since the calculation will happen later, in `replicate_to_all_cores/create_effective_replication_map`

gleb-cloudius replied:
But it still can be set in `update_pending_ranges` not as a separate call. Why I do not like the separate call is that, at least now, you need to know to call it before you call `update_pending_ranges` . This is not obvious and not clear.

gusev-p replied:
ok

gusev-p commented on locator/token_metadata.cc:
```diff
@@ -773,122 +710,107 @@ token_metadata_impl::has_pending_ranges(sstring keyspace_name, inet_address endp
         return make_ready_future<>();
     }
 
     return async([this, &unpimplified_this, &strategy, keyspace_name, &get_dc_rack] () mutable {
```
Yes, we can, but it's deferred to the follow-up, where `update_pending_ranges` is replaced with another function `update_topology_change_info`, which calculates only common data, which is required for all replication strategies. And `update_topology_change_info` is made a co-routune.

gusev-p commented on locator/token_metadata.cc:
```diff
@@ -187,28 +186,28 @@ class token_metadata_impl final {
-    future<token_metadata_impl> clone_after_all_left() const noexcept {
-      return clone_only_token_map(false).then([this] (token_metadata_impl all_left_metadata) {
+    future<std::unique_ptr<token_metadata_impl>> clone_after_all_left() const noexcept {
+      return clone_only_token_map(false).then([this] (std::unique_ptr<token_metadata_impl> all_left_metadata) {
```
fixed

gusev-p commented on locator/token_metadata.cc:
```diff
@@ -226,51 +227,22 @@ class token_metadata_impl final {
+    *
+    * We construct an updated version of the token_metadata by incorporating
+    * all proposed modifications (join, bootstrap, and replace operations).
+    * Subsequently, for each token, we compare the outcomes of the calculate_natural_endpoints
```
fixed

gusev-p commented:
v7:
  * small review fixes

gleb-cloudius commented on locator/token_metadata.cc:
```diff
@@ -773,122 +710,107 @@ token_metadata_impl::has_pending_ranges(sstring keyspace_name, inet_address endp
+            std::copy(begin(sorted_tokens()), end(sorted_tokens()), begin(tokens));
+            for (const auto& p: get_bootstrap_tokens()) {
+                tokens.push_back(p.first);
+            }
```
I do not insist of course. You are not the only one who dislikes boost (I do not mind it), but you do use std::copy() above instead of an open loop (which is good). Using an algorithm is better than open code a loop because it clearly shows the intention of the code without a need to read the loop. IIRC we do not use c++20 ranges because clang does not support it this is why we have to use boost sometimes.

avikivity replied:
boost::join typically generates terrible code because it needs to choose between several iterator types on each iteration. Separate loops don't make any choices.

gleb-cloudius replied:
I did say it will not be more efficient :)

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -356,13 +356,28 @@ future<> storage_service::topology_state_load(cdc::generation_service& cdc_gen_s
                           id, ip, rs.state, rs.datacenter, rs.rack, rs.ring->state, rs.ring->tokens);
 
+            {
+                const auto read_new = rs.ring.value().state == ring_slice::replication_state::write_both_read_new;
```
Don't you need to rebase it on master? Kamil moved state out to be a global per topology, bot per node.

gusev-p replied:
done

gusev-p commented:
v8:

rebased on master, use `_topology_state_machine._topology.tstate` instead of per node state.

avikivity commented on test/boost/token_metadata_test.cc:
```diff
@@ -0,0 +1,191 @@
+    }
+}
+
+SEASTAR_THREAD_TEST_CASE(test_pending_endpoints_for_bootstrap_first_node) {
```
I don't like unit test files with ~200 lines of code. We'll compile hundreds of thousands of headers file lines just to emit these.

No need to change. I propose we do a consolidation pass later where we merge related small files and split overgrown large files.

tgrabiec replied:
There's locator_topology_test, which is a candidate for unification with this.

mykaul replied:
OTOH, I'm not sure - does this test enough this code?

gusev-p replied:
> There's locator_topology_test, which is a candidate for unification with this.

I'll create a follow-up PR with the merge.

gusev-p replied:
> OTOH, I'm not sure - does this test enough this code?

These tests cover only aspects of token_metadata important for the refactoring, not the entire scope.

avikivity commented on locator/token_metadata.cc:
```diff
@@ -226,51 +227,22 @@ class token_metadata_impl final {
+    * function applied to both the previous and the new token_metadata.
+    * Endpoints present in the updated version but absent in the original one
+    * ought to be appended to the pending_ranges.
+    */
```
That's super nice.

avikivity commented on locator/token_metadata.cc:
```diff
@@ -773,122 +710,107 @@ token_metadata_impl::has_pending_ranges(sstring keyspace_name, inet_address endp
-        // At this stage newPendingRanges has been updated according to leaving and bootstrapping nodes.
-        set_pending_ranges(keyspace_name, std::move(new_pending_ranges), can_yield::yes);
+        // new_token_metadata incorporates all the changes from leaving, bootstrapping and replacing
+        const auto new_token_metadata = token_metadata([&]() -> std::unique_ptr<token_metadata_impl> {
```
Style: I prefer immediately-invoked lambdas to be called with std::invoke rather than the function call operator. The reason is that introducing the lambda creates the expectation that it isn't immediately called (as that's how most lambdas are used), and then the expectation is shattered when the call operator is seen. Usually one has to re-read the code with the new knowledge. With std::invoke() you read the code knowing it's going to run now.

No need to change for now, just keep in mind for the future.

avikivity commented on locator/token_metadata.cc:
```diff
@@ -773,122 +710,107 @@ token_metadata_impl::has_pending_ranges(sstring keyspace_name, inet_address endp
+                for (const auto& e: new_endpoints) {
+                    if (!old_endpoints.contains(e)) {
+                        pending_endpoints.insert(e);
+                    }
```
Can insert unconditionally, since insert is idempotent.

gusev-p replied:
we need `pending_endpoints = new_endpoints.subtract(old_endpoints)`, can't do this unconditionally.

avikivity replied:
I didn't even see that the set being tested isn't the set being updated.

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1020/

bhalevy commented on locator/token_metadata.cc:
```diff
@@ -774,122 +734,114 @@ token_metadata_impl::has_pending_ranges(sstring keyspace_name, inet_address endp
         const token_metadata& unpimplified_this,
-        const abstract_replication_strategy& strategy, const sstring& keyspace_name, dc_rack_fn& get_dc_rack) {
+        const abstract_replication_strategy& strategy, const sstring& keyspace_name, dc_rack_fn& get_dc_rack,
+        bool request_read_new) {
```
nit: coding style violation.
new param should either be added to the previous line (I'd prefer that) or start with double-indent.

avikivity replied:
All the parameters are double-indented already.

tgrabiec commented on locator/token_metadata.cc:
```diff
@@ -60,7 +60,19 @@ class token_metadata_impl final {
+    // For each keyspace, migration_info contains ranges of tokens and
+    // corresponding replicas to which writes or reads will be directed:
+    //   - pending_endpoints - will be appended to normal endpoints for writes;
+    //   - read_endpoints - will completely replace normal endpoints for reads.
```
Why do we need to distinguish read_endpoints?

tgrabiec replied:
read_endpoints is not set in this commit. If it's set later, it would be better to introduce it then.

gusev-p replied:
> Why do we need to distinguish read_endpoints?

`pending_endpoints` are set at `write_both_read_old`, they can only be appended to the normal endpoints, and they have some usages besides just being appended to normal endpoints: https://github.com/scylladb/scylladb/blob/master/service/storage_proxy.cc#L1221, 
https://github.com/scylladb/scylladb/blob/master/service/storage_proxy.cc#L1599
https://github.com/scylladb/scylladb/blob/master/service/storage_proxy.cc#L3626

`read_endpoints` on the other hand are set at `write_both_read_new` and they can skip some endpoints from `normal_endpoints`, that's why they completely replace them.

gusev-p replied:
> read_endpoints is not set in this commit. If it's set later, it would be better to introduce it then.

ok

tgrabiec replied:
> If it's set later, it would be better to introduce it then.

No need to do it now, just for the future.

gusev-p replied:
> No need to do it now, just for the future.

already updated

tgrabiec commented on service/storage_service.hh:
```diff
@@ -190,8 +190,8 @@ private:
-    future<> update_pending_ranges(mutable_token_metadata_ptr tmptr, sstring reason);
-    future<> update_pending_ranges(sstring reason, acquire_merge_lock aml = acquire_merge_lock::yes);
+    future<> update_pending_ranges(mutable_token_metadata_ptr tmptr, sstring reason, bool request_read_new = false);
+    future<> update_pending_ranges(sstring reason, acquire_merge_lock aml = acquire_merge_lock::yes, bool request_read_new = false);
```
This default seems unsafe, because a later update_pending_ranges() which misses the current value can wipe out request_read_new.

gusev-p replied:
ok

gusev-p replied:
Ah, I confused it with `token_metadata::update_pending_ranges`, I have no default for it there.

I don't want to remove default here:
* there are a lot of usages for this `update_pending_ranges`, in all of them this parameter is irrelevant and it only will make a noise if we state false explicitly. It only matters when transitioning to `write_both_read_new`, it's 3 usages out of total of two dozens over `storage_service`.
* we will have to move `request_read_new` before `acquire_merge_lock` for the second overload, this will make unnecessary changes

tgrabiec replied:
> It only matters when transitioning to write_both_read_new

But once we transition to write_both_read_new, shouldn't request_read_new=true stick to the token metadata until revoked, even in later transitions? There will be more transitions after the transition to write_both_read_new. 

Maybe it should be a flag in token metadata which update_pending_ranges() picks up. This way we also avoid modifying places which don't care about it.

gusev-p replied:
Now I see your point, and it's the argument I missed when @gleb-cloudius convinced me to do the opposite, initially there was a separate method to set this flag.

@gleb-cloudius are those numerous gossiper-induced `update_pending_ranges` supposed to be called when raft topology changes is enabled? should we explicitly disable them?

gleb-cloudius replied:
They should not be! We need to assert if they do.

gusev-p commented on locator/token_metadata.cc:
```diff
@@ -773,122 +710,107 @@ token_metadata_impl::has_pending_ranges(sstring keyspace_name, inet_address endp
+            for (auto& [endpoint, tokens]: new_normal_tokens) {
+                result->update_topology(endpoint,
+                    get_dc_rack(endpoint),
+                    _normal_token_owners.contains(endpoint) ? node::state::normal : node::state::joining);
```
I don' think it affects anything, `new_token_metadata` has quite restricted usage, but from general point of view you're probably right, will update it.

gusev-p commented on locator/token_metadata.cc:
```diff
@@ -773,122 +710,107 @@ token_metadata_impl::has_pending_ranges(sstring keyspace_name, inet_address endp
+                    }
+                    new_normal_tokens[it->second].insert(token);
+                }
+                for (const auto& [replace_from, replace_to]: _replacing_endpoints) {
```
what's the point of replacing a node with itself? should it be allowed in the first place?

bhalevy replied:
It's entirely possible to replace a node that is down with a new node having the same ip address.
Fortunately, we changed it so the new node will have a new host_id. In the past it even inherited the replaced node's host_id which was even harder.

gusev-p replied:
ok, git it, thanks

tgrabiec replied:
> tokens for replaced node will be added back in the loop "apply new_normal_tokens"

Right.

> the behaviour is changed though, previously pending_endpoints included the replaced endpoint, after the change it doesn't, this is demonstrated by the test. I'm not sure, is this important?

In the case of replacing with the same IP, you mean?

I think we should preserve the old behavior, because otherwise replacing node will not get the writes it's supposed to. The coordinator does this:

```
storage_proxy::create_write_response_handler_helper(...) {
    inet_address_vector_replica_set natural_endpoints = erm->get_natural_endpoints_without_node_being_replaced(token);
    inet_address_vector_topology_change pending_endpoints = erm->get_pending_endpoints(token, s->ks_name());
```

get_natural_endpoints_without_node_being_replaced() will filter-out the node which is being replaced. So the replacing node must be in pending endpoints for it to receive the writes.

bhalevy commented:
@gusev-p any news on the system test results (https://github.com/scylladb/scylladb/pull/13376#issuecomment-1534415380)?

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1092/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1096/

gusev-p commented:
v9:
  * fixed @tgrabiec comments: 
    * separate method to switch to read_new
    * special handling for replace(ep, ep)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1111/

gusev-p commented:
> @gusev-p any news on the system test results ([#13376 (comment)](https://github.com/scylladb/scylladb/pull/13376#issuecomment-1534415380))?

@bhalevy 

All tests have passed when running them all in parallel, except the two:
* **update_cluster_layout_tests.py::TestUpdateClusterLayout::test_simple_add_two_nodes_in_parallel** passed on master and on branch when run alone
* **update_cluster_layout_tests.py::TestUpdateClusterLayout::test_decommission_node_while_gossip_partly_blocked** failed both on master and on branch when run alone.

So I don't think the changes from this PR will make anything worse.

tgrabiec commented:
I need re-review from @gleb-cloudius

gusev-p commented:
v10:
  * fixed commits splitting, unit tests accidentally merged with refactoring

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1134/

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -354,6 +354,13 @@ future<> storage_service::topology_state_load(cdc::generation_service& cdc_gen_s
+            const auto stage = _topology_state_machine._topology.tstate == topology::transition_state::write_both_read_new
+                ? locator::topology_change_stage::read_new
+                : locator::topology_change_stage::read_old;
+            tmptr->set_topology_change_stage(stage);
```
Here it will simply be `tmptr->set_topology_change_stage(_topology_state_machine._topology.tstate)`. At some point we will need to merge `topology_metadata` and `topology` classes.

kbr-scylla replied:
> At some point we will need to merge topology_metadata and topology classes?

Why?

To me it makes as much sense as merging `gossiper::_endpoint_state_map` with `token_metadata`.

gleb-cloudius replied:
> Why?

They largely duplicate one another. Topology has some data that tm does not, but it can be stored in new `node` structure now. Well, may be what we really need to unify is `service::topology` and `locator::topology` but the `token_metadata` is closely coupled with the later anyway.

gusev-p commented on locator/token_metadata.hh:
```diff
@@ -70,6 +70,16 @@ struct host_id_or_endpoint {
+//     read_old - no read_endpoints will be stored in update_pending_ranges,
+//                all reads are directed to normal endpoints;
+//     read_new - initiates update_pending_ranges to calculate and store new ranges for read requests.
+enum class topology_change_stage {
```
This will lead to the refactoring of `service::topology`, creating a coupling with `locator::topology` just to reuse this enum. This doesn't seem worth it to me.

Maybe we should unite these two topologies (I'm not sure about that), but it sounds like a separate task with its own objective and value.

gusev-p replied:
updated

avikivity commented on locator/token_metadata.hh:
```diff
@@ -267,6 +268,16 @@ public:
+    // This function returns a list of nodes to which a read request should be directed.
+    // Returns not null only during topology changes, if _topology_change_stage == read_new and
+    // new set of replicas differs from the old one.
+    std::optional<inet_address_vector_replica_set> endpoints_for_reading(const token& token, const sstring& keyspace_name) const;
```
I think it would be better to use it unconditionally. But okay to discuss it post-merge.

gusev-p replied:
This would mean complete duplication of the `replication_map` in vnode_erm. Here I tried to be memory efficient (store only ranges with diffs only during `write_both_read_new`) and do it similarly to `pending_ranges` for writes to reduce the scope of changes.

`replication_map` in `vnode_erm` seems more efficient than these interval trees - it's just a sorted vector + has_map. In [this follow-up PR](https://github.com/scylladb/scylladb/pull/13715) I move these interval trees into erm. We could possibly go further and turn them into hash_maps: vnode -> pending_endpoints/read_endpoints. We can't just merge them into natural_endpoints since each `natural_endpoints`, `pending_endpoints` and `pending_endpoints` have their own distinct usages.

avikivity replied:
Yes. How about this:

1. Wait for @bhalevy to switch these to use node* (or uint16_t index to a node vector) instead of inet_address_vector_blah_blah_blah.
2. Store those ranges contiguously as sub-spans of a large vector (to avoid allocating an object per range)
3. change the APIs to return std::span\<node_index_t>

gusev-p replied:
https://github.com/scylladb/scylladb/issues/13839

avikivity commented on service/storage_proxy.cc:
```diff
@@ -6006,9 +6006,14 @@ void storage_proxy::sort_endpoints_by_proximity(const locator::topology& topo, i
+    auto endpoints = erm.get_token_metadata_ptr()->endpoints_for_reading(token, ks_name);
+    if (!endpoints) {
+        endpoints = erm.get_natural_endpoints_without_node_being_replaced(token);
+    }
```
e.g. here. Have just one call.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -6006,9 +6006,14 @@ void storage_proxy::sort_endpoints_by_proximity(const locator::topology& topo, i
+    }
+    auto it = boost::range::remove_if(*endpoints, std::not_fn(std::bind_front(&storage_proxy::is_alive, this)));
+    endpoints->erase(it, endpoints->end());
+    sort_endpoints_by_proximity(erm.get_topology(), *endpoints);
```
We should pre-sort in effective_replication_map (but not in this series).

gusev-p replied:
https://github.com/scylladb/scylladb/issues/13829

kbr-scylla commented:
Queued.
Good stuff.

gleb-cloudius commented on locator/token_metadata.cc:
```diff
@@ -851,6 +862,14 @@ future<> token_metadata_impl::update_pending_ranges(
+
+                // in order not to waste memory, we update read_endpoints only if the
+                // new endpoints differs from the old one
+                if (_topology_transition_state == service::topology::transition_state::write_both_read_new &&
```
This check was in the different place in the previous version, so if such a state was added the test would have needed a fix anyway. The way to handle it in a more general way is to add `is_read_new_state(state&)` function into `service::topology` and still the one who adds a state will have to remember to update it.

gusev-p replied:
> The way to handle it in a more general way is to add is_read_new_state(state&) function into service::topology and still the one who adds a state will have to remember to update it.

this feels like a hacky way to say the same thing as in my prev version: just convert one set of states to another

gleb-cloudius replied:
> > both sets are describing topology transitions.
> 
> They are not: the enum in `token_metadata` described just how reads should be conducted, not all sm states. It was meant to be equivalent to `bool read_new`, just more type safe, like `bool_class`. Maybe it should have been just `bool_class<class read_new_tag>`.

Exactly. If that was your intention `bool_class` is the one you should have used. But I actually liked setting the state, just did not like having additional enum to map to. More flexible IMO. But yes, there is a chance we will not need the flexibility.

tgrabiec replied:
A way to protect against condition rotting in case we add more states is to have a switch on the state which causes compilation error until someone makes a decision how to handle the new state in this conversion.
